destination/clickhouse.go

```diff
@@ -30,146 +31,185 @@ type ClickHouseConnection struct {
+	}, conn.ctx, "DescribeTable")
 }
 
 func (conn *ClickHouseConnection) CreateTable(schemaName string, tableName string, tableDescription *TableDescription) error {
-	statement, err := GetCreateTableStatement(schemaName, tableName, tableDescription)
-	if err != nil {
-		return err
-	}
-	if err := conn.Exec(conn.ctx, statement); err != nil {
-		return fmt.Errorf("error while executing %s: %w", statement, err)
-	}
-	return nil
+	return RetryNetError(func() error {
+		statement, err := GetCreateTableStatement(schemaName, tableName, tableDescription)
+		if err != nil {
+			return err
+		}
+		if err = conn.Exec(conn.ctx, statement); err != nil {
+			LogError(fmt.Errorf("error while executing %s: %w", statement, err))
+			return err
+		}
+		return nil
+	}, conn.ctx, "CreateTable")
 }
 
 func (conn *ClickHouseConnection) AlterTable(schemaName string, tableName string, ops []*AlterTableOp) error {
-	statement, err := GetAlterTableStatement(schemaName, tableName, ops)
-	if err != nil {
-		return err
-	}
-	if err := conn.Exec(conn.ctx, statement); err != nil {
-		return fmt.Errorf("error while executing %s: %w", statement, err)
-	}
-	return nil
+	return RetryNetError(func() error {
+		statement, err := GetAlterTableStatement(schemaName, tableName, ops)
+		if err != nil {
+			return err
+		}
+		if err = conn.Exec(conn.ctx, statement); err != nil {
+			LogError(fmt.Errorf("error while executing %s: %w", statement, err))
+			return err
+		}
+		return nil
+	}, conn.ctx, "AlterTable")
 }
 
 func (conn *ClickHouseConnection) TruncateTable(schemaName string, tableName string) error {
-	statement, err := GetTruncateTableStatement(schemaName, tableName)
-	if err != nil {
-		return err
-	}
-	if err := conn.Exec(conn.ctx, statement); err != nil {
-		return fmt.Errorf("error while executing %s: %w", statement, err)
-	}
-	return nil
+	return RetryNetError(func() error {
+		statement, err := GetTruncateTableStatement(schemaName, tableName)
+		if err != nil {
+			return err
+		}
+		if err = conn.Exec(conn.ctx, statement); err != nil {
+			LogError(fmt.Errorf("error while executing %s: %w", statement, err))
+			return err
+		}
+		return nil
+	}, conn.ctx, "TruncateTable")
 }
 
 func (conn *ClickHouseConnection) InsertBatch(
 	fullTableName string,
 	rows [][]interface{},
 	skipIdx map[int]bool,
 	opType WriteBatchOpType,
+	async bool,
 ) error {
-	if len(skipIdx) == len(rows) {
-		LogWarn(fmt.Sprintf("[InsertBatch - %s] All rows are skipped for %s", opType, fullTableName))
-		return nil
-	}
-	batch, err := conn.PrepareBatch(conn.ctx, fmt.Sprintf("INSERT INTO %s", fullTableName))
-	if err != nil {
-		return fmt.Errorf("error while preparing a batch for %s: %w", fullTableName, err)
-	}
-	for i, row := range rows {
-		if skipIdx[i] {
-			continue
+	ctx := conn.ctx
+	if async {
+		ctx = clickhouse.Context(conn.ctx, clickhouse.WithSettings(clickhouse.Settings{
+			"async_insert":          1,
+			"wait_for_async_insert": 1,
+		}))
+	}
+	return RetryNetError(func() error {
```

Comment:
We should have a good motivation to retry on all network errors. Not sure what about duplicates in ReplacingMergeTree?
Should it be configurable and end-user should decide on it?

Reply:
We should be fine.

```sql
CREATE OR REPLACE TABLE foo
(
    `s` String,
    `ver` Int
)
ENGINE = ReplacingMergeTree(ver)
ORDER BY s

INSERT INTO foo VALUES ('foo', 1), ('foo', 1), ('bar', 1)
INSERT INTO foo VALUES ('foo', 1), ('foo', 1), ('bar', 1)

SELECT *
FROM foo
FINAL

┌─s───┬─ver─┐
│ bar │   1 │
│ foo │   1 │
└─────┴─────┘